### DISCUSSION TITLE
Combining Flet with a Python backend framework.

### DISCUSSION THREAD

**Discussion by joshajiniran** [flet-dev/flet#752] [Q&A]
I have been wondering how anyone would combine Flet and a Python web framework such as FastAPI, DRF, Flask etc. Currently, I have built a Django-Ninja API and would like to use Flet, for the interfaces.

My major concern though, is figuring out how to architect the codebase in terms of structure and also deployment (e.g I was thinking deploy the Flet app to [fly.io](https://fly.io) which would be consuming the API deployed to Render or DO etc.) Or is there a way to have Flet render the frontend parts of the app as Templates do and then have a single app deployment

I'd appreciate a guide on this.

**Comment by FeodorFitsner** [comment]
Currently, Flet is a server-side framework, but the whole point of FastAPI and such is to provide backend services to a frontend app via JSON. It would have more sense if there is an ability to host Flet app inside FastAPI and call its services directly without HTTP overhead. We are working on a "light" version of Flet server written in Python which, I believe, could be hosted inside FastAPI.

**Reply by joshajiniran** [reply]
Oh, great job, so I can talk directly to a DB for any operation, but is there an ORM like feature or guide.

Keep up the good work boss

**Reply by mnesarco** [reply]
Some experiments here: https://github.com/flet-dev/flet/issues/775#issuecomment-1375558743

**Comment by antivodonik** [comment]
I'm not sure I understand exactly what you're trying to achieve, but I played with Flet and FastAPI yesterday, and I found an interesting solution to dynamically render Flet controls from JSON data sent by the FastAPI server. Here's the example server code:

```
from fastapi import FastAPI

app = FastAPI()


@app.get("/")
def get_root():
    return {"content": [
                        {"type": "text", "data": {"value": "Hello World", "size": 30}},
                        {"type": "input", "data": {"hint_text": "Hint text", "width": 300}}
                       ]}
```

And the Flet client:

```
import flet as ft
import requests


def process_data(data_in: dict):
    rt = None
    d = data_in['data']
    if data_in['type'] == 'text':
        rt = ft.Text()
        for key in list(d.keys()):
            rt.__setattr__(key, d[key])
    elif data_in['type'] == 'input':
        rt = ft.TextField()
        for key in list(d.keys()):
            rt.__setattr__(key, d[key])
    return rt


def main(page: ft.Page):
    ntxt = requests.get('http://localhost:8000')
    jtxt = ntxt.json()
    if 'content' in jtxt.keys():
        for item in jtxt['content']:
            page.add(process_data(item))


ft.app(target=main)
```

Now, it depends on how you want to structure the JSON, but as long as you use dictionary key that is the same as a Flet control property name, this can work almost universally.
